File: app/webapp/static/audit.js
/* Audit tab (#12): per-run trace drill-down — the reason the radar is trustworthy.
 *
 * Read-only view over the persisted trace (#7). The run list shows each run's
 * mode (live/dry-run), parameters, and funnel; clicking a run drills into the
 * per-chat decision record: the input delta, the Stage-1 roots, the exact LLM
 * prompts, the raw response, the parsed verdict, the final action, and the
 * Telegram text. Resync/reprocess maintenance events are interleaved so the whole
 * timeline is visible. All chat names / message text / prompts go in via
 * textContent only — never innerHTML (privacy). */

import { els, state } from './state.js';
import { jsonApi } from './api.js';
import { fmtLocalDateTime, kindLabel, renderFunnelCells, renderSourceFunnels } from './format.js';

function auditState() { return state.audit; }

const FAMILY_KINDS = ['traffic-check', 'calendar-scan'];

function isFamilyKind(kind) { return FAMILY_KINDS.indexOf(kind) !== -1; }

// Kind filter (#163): 'messages' groups the message-pipeline kinds; the family
// checks filter individually.
function matchesKindFilter(run, filter) {
  if (!filter || filter === 'all') return true;
  if (filter === 'messages') return !isFamilyKind(run.kind);
  return run.kind === filter;
}

// Live = a real run that synced + (maybe) delivered; dry_run = replay on stored
// data; review = the legacy process-only verb. Badge keys map to CSS colors.
const MODE_META = {
  live: { label: 'Live', cls: 'live' },
  dry_run: { label: 'Dry run', cls: 'dry' },
  review: { label: 'Process', cls: 'review' },
};

function modeMeta(mode) { return MODE_META[mode] || { label: mode || '?', cls: 'review' }; }

function statusBadge(status) {
  if (status === 'completed') return { text: 'OK', cls: 'ok' };
  if (status === 'failed') return { text: 'KO', cls: 'ko' };
  if (status === 'running' || status === 'pending') return { text: '··', cls: 'run' };
  return { text: '··', cls: 'run' };
}

// The final per-chat verdict, mapped to a short label + tone for the trace header.
const ACTION_META = {
  actionable: { label: 'Actionable', cls: 'act' },
  not_actionable: { label: 'Not actionable', cls: 'muted' },
  contract_error: { label: 'Contract error', cls: 'err' },
  llm_truncated: { label: 'LLM truncated', cls: 'err' },
};

function actionMeta(action) { return ACTION_META[action] || { label: action || '?', cls: 'muted' }; }

// One compact line summarizing the run for the list: the funnel for message
// runs, the structured payload counts for family checks (#163).
function runSummaryLine(run) {
  if (isFamilyKind(run.kind)) {
    const s = run.summary || {};
    if (run.kind === 'traffic-check') {
      return `${(s.checked || []).length} checked · ${s.alerts || 0} alert(s)`;
    }
    return `${(s.conflicts || []).length} conflict(s) · `
      + `${(s.missing_locations || s.unknown_locations || []).length} missing location(s)`;
  }
  const f = run.funnel;
  if (!f) return '';
  return [
    `${f.messages_synced} synced`,
    `${f.chats_monitored} monitored`,
    ...(f.transcriptions ? [`${f.transcriptions} voice`] : []),
    `${f.stage1_passed} Stage 1`,
    `${f.stage2_llm_calls} LLM`,
    `${f.actionable} actionable`,
  ].join(' · ');
}

function paramsSummary(params) {
  if (!params || typeof params !== 'object') return '';
  const bits = [];
  for (const k of Object.keys(params)) {
    if (params[k] !== null && params[k] !== undefined) bits.push(`${k}=${params[k]}`);
  }
  return bits.join(' ');
}

// ----------------------------------------------------------- run list

function runListItem(run) {
  const li = document.createElement('li');
  li.className = 'audit-run-li';
  const ax = auditState();
  if (ax.selected === run.id) li.classList.add('selected');

  const top = document.createElement('div');
  top.className = 'audit-run-top';

  const sb = statusBadge(run.status);
  const status = document.createElement('span');
  status.className = 'exec-run-badge ' + sb.cls;
  status.textContent = sb.text;

  const name = document.createElement('span');
  name.className = 'exec-run-name small';
  name.textContent = kindLabel(run.kind);

  const mm = modeMeta(run.mode);
  const mode = document.createElement('span');
  mode.className = 'audit-mode-badge ' + mm.cls;
  mode.textContent = mm.label;

  const when = document.createElement('span');
  when.className = 'audit-run-when muted small';
  const params = paramsSummary(run.params);
  when.textContent = fmtLocalDateTime(run.started_at) + (params ? ' · ' + params : '');

  top.append(status, name, mode, when);

  const summary = document.createElement('p');
  summary.className = 'audit-run-funnel muted small';
  summary.textContent = runSummaryLine(run);

  li.append(top, summary);
  li.addEventListener('click', function () { selectRun(run.id); });
  return li;
}

// A resync/reprocess maintenance marker, visually distinct from review runs.
function syncListItem(sync) {
  const li = document.createElement('li');
  li.className = 'audit-sync-li';

  const tag = document.createElement('span');
  tag.className = 'audit-sync-tag';
  tag.textContent = sync.source === 'reprocess' ? 'Rebuild' : 'Resync';

  const when = document.createElement('span');
  when.className = 'audit-run-when muted small';
  when.textContent = fmtLocalDateTime(sync.ran_at);

  const delta = document.createElement('span');
  delta.className = 'audit-sync-delta muted small';
  const chatBit = sync.chats_added ? ` · +${sync.chats_added} chat${sync.chats_added > 1 ? 's' : ''}` : '';
  delta.textContent = `+${sync.messages_added} msg${sync.messages_added === 1 ? '' : 's'}${chatBit}`;

  li.append(tag, when, delta);
  return li;
}

function renderRuns() {
  const ax = auditState();
  els.auditRuns.textContent = '';
  const shown = ax.runs.filter(function (r) { return matchesKindFilter(r, ax.kindFilter); });
  els.auditRunsEmpty.hidden = shown.length > 0;
  for (const run of shown) els.auditRuns.appendChild(runListItem(run));
}

// ----------------------------------------------------------- run detail

function funnelCells(run) {
  const f = run.funnel || {};
  return [
    { label: 'Synced', value: f.messages_synced },
    { label: 'Monitored', value: f.chats_monitored },
    { label: 'Reviewed', value: f.chats_reviewed },
    { label: 'Transcribed', value: f.transcriptions },
    { label: 'Stage 1', value: f.stage1_passed },
    { label: 'LLM', value: f.stage2_llm_calls },
    { label: 'Actionable', value: f.actionable },
    { label: 'Notify', value: run.notification_status },
  ];
}

function renderFunnel(run) {
  renderFunnelCells(els.auditFunnel, funnelCells(run));
}

// A labelled <pre> block; skipped entirely when there's nothing to show so the
// trace stays compact. Text goes in via textContent (privacy + no markup eval).
function traceField(title, text) {
  if (text === null || text === undefined || text === '') return null;
  const wrap = document.createElement('div');
  wrap.className = 'audit-field';
  const h = document.createElement('h5');
  h.className = 'audit-field-title';
  h.textContent = title;
  const pre = document.createElement('pre');
  pre.className = 'codebox audit-pre';
  pre.textContent = typeof text === 'string' ? text : JSON.stringify(text, null, 2);
  wrap.append(h, pre);
  return wrap;
}

// One message inside a chat trace: its text plus why it did / didn't trigger.
// Stage-1 shows the exact keyword roots that message matched (or "no keyword");
// the LLM column is shown only when the LLM ran, marking whether the model
// flagged this message as evidence. Eliminates the "which ones triggered?" black
// box (#12). All content via textContent (privacy + no markup eval).
function messageRow(m, llmCalled, evidence) {
  const li = document.createElement('li');
  li.className = 'audit-msg';

  const head = document.createElement('div');
  head.className = 'audit-msg-head';

  const sender = document.createElement('span');
  sender.className = 'audit-msg-sender small';
  sender.textContent = m.sender || 'unknown';
  head.appendChild(sender);

  // A voice badge marks a note whose text is the transcription fed into
  // analysis (#36).
  if (m.type === 'voice') {
    const voice = document.createElement('span');
    voice.className = 'audit-msg-badge muted';
    voice.textContent = 'voice';
    head.appendChild(voice);
  }

  const roots = Array.isArray(m.roots) ? m.roots : [];
  const buckets = Array.isArray(m.buckets) ? m.buckets : [];
  if (buckets.length) {
    const bucket = document.createElement('span');
    bucket.className = 'audit-msg-badge act';
    bucket.textContent = buckets.join(', ');
    head.appendChild(bucket);
  }
  const s1 = document.createElement('span');
  s1.className = 'audit-msg-badge ' + (roots.length ? 'act' : 'muted');
  s1.textContent = roots.length ? roots.join(', ') : 'no keyword';
  head.appendChild(s1);

  if (llmCalled) {
    const flagged = evidence.some(function (e) { return String(e) === String(m.id); });
    const s2 = document.createElement('span');
    s2.className = 'audit-msg-badge ' + (flagged ? 'act' : 'muted');
    s2.textContent = flagged ? 'LLM flagged' : 'LLM: not flagged';
    head.appendChild(s2);
  }

  const text = document.createElement('p');
  text.className = 'audit-msg-text';
  text.textContent = m.text || '(no text)';

  li.append(head, text);
  return li;
}

function messagesList(t) {
  const msgs = Array.isArray(t.messages) ? t.messages : [];
  if (!msgs.length) return null;
  const evidence = (t.parsed_result && Array.isArray(t.parsed_result.evidence_message_ids))
    ? t.parsed_result.evidence_message_ids : [];
  const wrap = document.createElement('div');
  wrap.className = 'audit-field';
  const h = document.createElement('h5');
  h.className = 'audit-field-title';
  h.textContent = `Messages (${msgs.length})`;
  const ul = document.createElement('ul');
  ul.className = 'audit-msg-list';
  for (const m of msgs) ul.appendChild(messageRow(m, t.llm_called, evidence));
  wrap.append(h, ul);
  return wrap;
}

function traceBlock(t) {
  // A vendored disclosure card (card--collapsible): title + verdict in the
  // summary's main cluster, chevron pinned right.
  const det = document.createElement('details');
  det.className = 'audit-trace card card--collapsible';

  const sum = document.createElement('summary');
  sum.className = 'collapse-summary';
  const main = document.createElement('span');
  main.className = 'collapse-main';
  const name = document.createElement('h3');
  name.className = 'collapse-title';
  name.textContent = t.display_name;
  const am = actionMeta(t.final_action);
  const verdict = document.createElement('span');
  verdict.className = 'audit-verdict ' + am.cls;
  verdict.textContent = am.label;
  main.append(name, verdict);
  const chevron = document.createElement('span');
  chevron.className = 'collapse-chevron';
  chevron.setAttribute('aria-hidden', 'true');
  chevron.textContent = '›';
  sum.append(main, chevron);
  det.appendChild(sum);

  const body = document.createElement('div');
  body.className = 'collapse-body audit-trace-body';

  // Stage progress line: did it pass Stage 1, was the LLM called?
  const stages = document.createElement('p');
  stages.className = 'muted small';
  const source = t.source === 'gmail' ? 'Gmail' : 'WhatsApp';
  const s1 = t.stage1_passed ? 'Stage 1 passed' : 'Stage 1 filtered';
  const s2 = t.llm_called ? 'LLM called' : 'LLM skipped';
  stages.textContent = `${source} · ${s1} · ${s2}`;
  body.appendChild(stages);

  // Per-message breakdown when the trace carries it (#12); older rows fall back
  // to the rendered input blob so historical traces still render.
  const perMessage = messagesList(t);
  const roots = Array.isArray(t.stage1_roots) ? t.stage1_roots.join(', ') : t.stage1_roots;
  const buckets = Array.isArray(t.stage1_buckets)
    ? t.stage1_buckets.join(', ') : t.stage1_buckets;
  const fields = [
    perMessage || traceField('Input messages', t.input_text),
    traceField('Stage-1 buckets matched', buckets),
    traceField('Stage-1 roots triggered', roots),
    traceField('LLM system prompt', t.llm_system_prompt),
    traceField('LLM user prompt', t.llm_user_prompt),
    traceField('Raw LLM response', t.llm_raw_response),
    traceField('Parsed verdict', t.parsed_result),
    traceField('Telegram text', t.telegram_text),
    traceField('Error', t.error),
  ];
  for (const f of fields) if (f) body.appendChild(f);

  det.appendChild(body);
  return det;
}

// Family-check drill-down (#163): the run's structured payload IS the trace —
// show the headline counts as funnel cells and the full payload verbatim.
function renderFamilyDetail(run) {
  const s = run.summary || {};
  const cells = run.kind === 'traffic-check'
    ? [
      { label: 'Checked', value: (s.checked || []).length },
      { label: 'Alerts', value: s.alerts },
      { label: 'Status', value: s.status },
    ]
    : [
      { label: 'Conflicts', value: (s.conflicts || []).length },
      { label: 'Missing loc.', value: (s.missing_locations || s.unknown_locations || []).length },
      { label: 'Status', value: s.status },
    ];
  renderFunnelCells(els.auditFunnel, cells);
  renderSourceFunnels(els.auditSourceFunnel, {});
  els.auditTraces.textContent = '';
  // Calendar-sync runs carry a per-event decision trace + the sent summary
  // (#168) — render those readably above the raw payload.
  const summaryMsg = s.summary && s.summary.text
    ? traceField(`Summary message (${s.summary.status || 'unknown'})`, s.summary.text)
    : null;
  if (summaryMsg) els.auditTraces.appendChild(summaryMsg);
  const decisions = (s.decisions || []).map((d) => {
    const when = fmtLocalDateTime(d.start);
    const flags = [];
    if (d.assumed) flags.push('missing location — assumed home');
    if (d.commute) flags.push('commute');
    return `${d.person} · ${when} · "${d.event}" → ${d.kind}`
      + ` (${d.source})${flags.length ? ' [' + flags.join(', ') + ']' : ''}`;
  });
  const decisionsBlock = decisions.length
    ? traceField(`Event decisions (${decisions.length})`, decisions.join('\n'))
    : null;
  if (decisionsBlock) els.auditTraces.appendChild(decisionsBlock);
  // Live phone-position coverage judgment (#177) — derived values only, the
  // payload never carries coordinates.
  const live = (s.live_coverage || []).map((c) => {
    if (!c.assessed) {
      return `${c.person}: no live fix (${c.presence_status}) — `
        + `calendar inference stands for ${(c.windows || []).join(', ')}`;
    }
    const whereabouts = c.at_home ? 'at home' : `~${c.eta_min} min from home`;
    return `${c.person} ${whereabouts} → '${c.window}' `
      + `${c.feasible ? 'reachable' : 'AT RISK'}`
      + ` (margin ${c.margin_min} min, fix ${c.presence_age_min} min old)`;
  });
  const liveBlock = live.length
    ? traceField(`Live coverage (${live.length})`, live.join('\n'))
    : null;
  if (liveBlock) els.auditTraces.appendChild(liveBlock);
  const payload = traceField('Run payload', run.summary);
  if (payload) els.auditTraces.appendChild(payload);
  els.auditTracesEmpty.hidden = !!payload;
  if (!payload) els.auditTracesEmpty.textContent = 'No payload recorded for this run.';
}

function renderDetail(data) {
  const run = data.run;
  els.auditDetailCard.hidden = false;
  const mm = modeMeta(run.mode);
  const kindBit = isFamilyKind(run.kind) ? kindLabel(run.kind) + ' — ' : '';
  // "Live run #40" / "Dry run #3" — don't double the word when the mode label
  // already ends in "run".
  els.auditDetailTitle.textContent = kindBit + (mm.label.toLowerCase().endsWith('run')
    ? `${mm.label} #${run.id}`
    : `${mm.label} run #${run.id}`);

  const bits = [run.status];
  if (run.started_at) bits.push('started ' + fmtLocalDateTime(run.started_at));
  const params = paramsSummary(run.params);
  if (params) bits.push(params);
  if (run.error) bits.push('error: ' + run.error);
  els.auditDetailMeta.textContent = bits.join(' · ');

  if (isFamilyKind(run.kind)) {
    renderFamilyDetail(run);
    els.auditDetailCard.scrollIntoView({ behavior: 'smooth', block: 'start' });
    return;
  }

  renderFunnel(run);
  renderSourceFunnels(els.auditSourceFunnel, run.sources || {});

  els.auditTraces.textContent = '';
  const traces = data.traces || [];
  els.auditTracesEmpty.hidden = traces.length > 0;
  // Reconcile the funnel when there's nothing to drill into: a live scan can sync
  // messages yet trace nothing because none landed in a monitored chat. Say so
  // explicitly rather than a bare "no trace" — without surfacing the unmonitored
  // chats' content (scope stays monitored-only, #12).
  if (!traces.length) {
    const synced = (run.funnel && run.funnel.messages_synced) || 0;
    els.auditTracesEmpty.textContent = synced
      ? `${synced} message${synced === 1 ? '' : 's'} synced, but none in a monitored chat — nothing to analyze.`
      : 'No per-chat trace recorded for this run.';
  }
  for (const t of traces) els.auditTraces.appendChild(traceBlock(t));

  // Bring the detail into view on a phone after tapping a run up the list.
  els.auditDetailCard.scrollIntoView({ behavior: 'smooth', block: 'start' });
}

async function selectRun(runId) {
  const ax = auditState();
  ax.selected = runId;
  renderRuns();
  try {
    const data = await jsonApi(`/api/audit/runs/${runId}`);
    ax.detail = data;
    renderDetail(data);
  } catch (_) {
    /* transient — re-tapping the run retries */
  }
}

function closeDetail() {
  const ax = auditState();
  ax.selected = null;
  ax.detail = null;
  els.auditDetailCard.hidden = true;
  renderRuns();
}

// ----------------------------------------------------------- fetch + wire

export async function fetchAudit() {
  let data;
  try {
    data = await jsonApi('/api/audit/runs');
  } catch (_) {
    return;  // 401 flips the login overlay; stay quiet otherwise.
  }
  const ax = auditState();
  ax.runs = data.runs || [];
  ax.syncs = data.syncs || [];
  renderRuns();

  // If a previously-selected run is gone (e.g. after a reprocess reset), drop it.
  if (ax.selected && !ax.runs.some(function (r) { return r.id === ax.selected; })) {
    closeDetail();
  }
}

export function wireAudit() {
  els.auditDetailClose.addEventListener('click', closeDetail);
  els.auditDetailCard.hidden = true;
  // Kind filter chips (#163): client-side filter over the fetched run list.
  els.auditKindFilter.addEventListener('click', function (ev) {
    const btn = ev.target.closest('.range-tab');
    if (!btn) return;
    auditState().kindFilter = btn.dataset.kind || 'all';
    for (const b of els.auditKindFilter.querySelectorAll('.range-tab')) {
      b.classList.toggle('active', b === btn);
    }
    renderRuns();
  });
}

File: app/webapp/routers/audit.py
"""Audit tab (#12): per-run trace drill-down — the surface that makes the radar
trustworthy.

A missed important message is a real failure, so for any run the operator must be
able to answer *why* each message was or wasn't promoted. After #7 every run
persists to ``review_runs`` (mode, params, funnel counters) and one
``analysis_trace`` row per (run, chat) capturing the full decision record — input
delta, Stage-1 keyword roots, the exact LLM system+user prompts, the raw model
response, the parsed verdict, the final action, and the Telegram text.

This router is **read-only** (SELECT only, no writes, no cursor changes) so it is
safe on the request path. It exposes a run list (with the funnel), a bounded
cross-run list of filtered decisions, and a per-run drill-down (the per-chat
trace). It never triggers scans — that is the Execution tab (#11).
"""

from __future__ import annotations

import json
import sqlite3
from datetime import UTC, datetime, timedelta
from math import ceil
from typing import Any

from fastapi import APIRouter, Depends, HTTPException

from app.webapp.routers._helpers import get_conn
from src.db import store

router = APIRouter()

# sync_log sources that represent data-maintenance runs (not review/scan runs).
# Surfaced in the audit timeline so resync/reprocess are visible alongside runs;
# 'scan'-sourced syncs are omitted because scans already appear as review_runs.
_MAINTENANCE_SOURCES = {"resync", "reprocess"}


def _loads(value: Any) -> Any:
    """Parse a stored JSON column, tolerating null/blank/legacy non-JSON text."""
    if value is None or value == "":
        return None
    if not isinstance(value, str):
        return value
    try:
        return json.loads(value)
    except (ValueError, TypeError):
        return value


def _run_list_row(row: sqlite3.Row) -> dict[str, Any]:
    """Shape a run row for the run list: identity + parsed params + funnel.

    Since #163 the table holds every run kind; family checks (traffic-check /
    calendar-scan) carry their structured payload in ``summary`` instead of a
    meaningful funnel.
    """
    return {
        "id": int(row["id"]),
        "kind": row["kind"],
        "summary": _loads(row["summary_json"]),
        "mode": row["mode"],
        "status": row["status"],
        "params": _loads(row["params_json"]),
        "started_at": row["started_at"],
        "completed_at": row["completed_at"],
        "notification_status": row["notification_status"],
        "error": row["error"],
        "sources": _loads(row["source_funnel_json"]) or {},
        "funnel": {
            "chats_synced": int(row["chats_synced"]),
            "messages_synced": int(row["messages_synced"]),
            "chats_monitored": int(row["chats_monitored"]),
            "chats_reviewed": int(row["chats_reviewed"]),
            "transcriptions": int(row["transcriptions"]),
            "stage1_passed": int(row["stage1_passed"]),
            "stage2_llm_calls": int(row["stage2_llm_calls"]),
            "actionable": int(row["actionable"]),
        },
    }


def _coverage_gap(
    offline_runs: list[sqlite3.Row], recovery: sqlite3.Row | None
) -> dict[str, Any] | None:
    """Collapse a multi-run connector outage into one Audit timeline marker."""
    if len(offline_runs) < 2:
        return None
    first = offline_runs[0]
    last = offline_runs[-1]
    elapsed = datetime.fromisoformat(last["started_at"]) - datetime.fromisoformat(
        first["started_at"]
    )
    return {
        "started_at": first["started_at"],
        "ended_at": last["started_at"],
        "duration_days": max(1, ceil(elapsed.total_seconds() / 86_400)),
        "failed_runs": len(offline_runs),
        "run_ids": [int(row["id"]) for row in offline_runs],
        "recovered_at": recovery["started_at"] if recovery is not None else None,
        "recovery_run_id": int(recovery["id"]) if recovery is not None else None,
    }


def _coverage_gaps(rows: list[sqlite3.Row]) -> list[dict[str, Any]]:
    """Return contiguous multi-run offline windows, newest first."""
    gaps: list[dict[str, Any]] = []
    offline_runs: list[sqlite3.Row] = []
    for row in rows:
        offline = row["status"] == "failed" and row["notification_status"] == "offline"
        if offline:
            offline_runs.append(row)
            continue
        gap = _coverage_gap(offline_runs, row)
        if gap is not None:
            gaps.append(gap)
        offline_runs = []
    gap = _coverage_gap(offline_runs, None)
    if gap is not None:
        gaps.append(gap)
    return list(reversed(gaps))


def _trace_row(row: sqlite3.Row) -> dict[str, Any]:
    """Shape one analysis_trace row (joined to chat name) into the full decision record."""
    return {
        "chat_id": int(row["chat_id"]),
        "source": row["source"],
        "display_name": row["display_name"],
        "input_message_ids": _loads(row["input_message_ids_json"]),
        "input_text": row["input_text"],
        "messages": _loads(row["messages_json"]),
        "stage1_passed": bool(row["stage1_passed"]),
        "stage1_roots": _loads(row["stage1_roots_json"]),
        "stage1_buckets": _loads(row["stage1_buckets_json"]),
        "llm_called": bool(row["llm_called"]),
        "llm_system_prompt": row["llm_system_prompt"],
        "llm_user_prompt": row["llm_user_prompt"],
        "llm_raw_response": row["llm_raw_response"],
        "parsed_result": _loads(row["parsed_result_json"]),
        "final_action": row["final_action"],
        "telegram_text": row["telegram_text"],
        "error": row["error"],
    }


def _filtered_trace_row(row: sqlite3.Row) -> dict[str, Any]:
    """Shape one cross-run filtered item without duplicating the full trace."""
    return {
        "trace_id": int(row["trace_id"]),
        "run_id": int(row["run_id"]),
        "created_at": row["created_at"],
        "source": row["source"],
        "display_name": row["display_name"],
        "stage1_passed": bool(row["stage1_passed"]),
        "stage1_roots": _loads(row["stage1_roots_json"]),
        "llm_called": bool(row["llm_called"]),
        "parsed_result": _loads(row["parsed_result_json"]),
        "final_action": row["final_action"],
    }


@router.get("/api/audit/runs")
async def list_runs(
    limit: int = 50,
    kind: str | None = None,
    conn: sqlite3.Connection = Depends(get_conn),
) -> dict[str, Any]:
    """Recent runs of every kind (with funnel/summary) plus maintenance markers.

    ``runs`` are the inspectable runs, newest first — message scans, process
    runs, and the family checks alike (#163); ``kind`` filters to one kind.
    ``syncs`` are the resync/reprocess data-maintenance events so they're
    visible in the same audit timeline (read-only, no schema beyond sync_log).
    """
    limit = max(1, min(limit, 200))
    runs = [
        r
        for r in (_run_list_row(row) for row in store.list_review_runs(conn, limit))
        if kind is None or r["kind"] == kind
    ]
    syncs = [
        {
            "ran_at": s["ran_at"],
            "source": s["source"],
            "chats_added": int(s["chats_added"]),
            "chats_updated": int(s["chats_updated"]),
            "messages_added": int(s["messages_added"]),
        }
        for s in store.recent_syncs(conn, limit)
        if s["source"] in _MAINTENANCE_SOURCES
    ]
    coverage_gaps = _coverage_gaps(store.list_live_scan_runs(conn))
    return {"runs": runs, "syncs": syncs, "coverage_gaps": coverage_gaps}


@router.get("/api/audit/filtered")
async def list_filtered(
    days: int = 30,
    limit: int = 50,
    offset: int = 0,
    conn: sqlite3.Connection = Depends(get_conn),
) -> dict[str, Any]:
    """Recent cross-run decisions that did not produce an actionable item."""
    days = max(1, min(days, 3650))
    limit = max(1, min(limit, 100))
    offset = max(0, offset)
    since = (datetime.now(UTC) - timedelta(days=days)).isoformat(timespec="seconds")
    total = store.count_filtered_traces(conn, since)
    items = [
        _filtered_trace_row(row)
        for row in store.list_filtered_traces(conn, since, limit, offset)
    ]
    return {
        "days": days,
        "limit": limit,
        "offset": offset,
        "total": total,
        "has_more": offset + len(items) < total,
        "items": items,
    }


@router.get("/api/audit/runs/{run_id}")
async def get_run(
    run_id: int,
    conn: sqlite3.Connection = Depends(get_conn),
) -> dict[str, Any]:
    """One run's header + funnel and its per-chat audit trace (the drill-down)."""
    run = store.review_run(conn, run_id)
    if run is None:
        raise HTTPException(status_code=404, detail="run not found")
    traces = [_trace_row(t) for t in store.traces_for_run(conn, run_id)]
    return {"run": _run_list_row(run), "traces": traces}
